cli/azd/pkg/environment/manager.go

```diff
@@ -0,0 +1,351 @@
+			fmt.Fprintf(
+				m.console.Handles().Stdout,
+				"environment name '%s' is invalid (it should contain only alphanumeric characters and hyphens)\n",
+				environmentName)
+			return nil, false, fmt.Errorf(
+				"environment name '%s' is invalid (it should contain only alphanumeric characters and hyphens)",
+				environmentName)
+		}
+
+		spec := &Spec{
+			Name: environmentName,
+		}
+
+		if err := m.ensureValidEnvironmentName(ctx, spec); err != nil {
+			return nil, false, err
+		}
+
+		return New(spec.Name), true, nil
+	}
+
+	env, isNew, err := loadOrCreateEnvironment()
+	switch {
+	case errors.Is(err, ErrNotFound):
+		return nil, fmt.Errorf("environment %s does not exist", environmentName)
+	case err != nil:
+		return nil, err
+	}
+
+	if isNew {
+		if env.GetEnvName() == "" {
+			env.SetEnvName(environmentName)
+		}
+
+		if err := m.Save(ctx, env); err != nil {
+			return nil, err
+		}
+
+		if err := m.azdContext.SetDefaultEnvironmentName(env.GetEnvName()); err != nil {
+			return nil, fmt.Errorf("saving default environment: %w", err)
+		}
+	}
+
+	return env, nil
+}
+
+// ConfigPath returns the path to the environment config file
+func (m *manager) ConfigPath(env *Environment) string {
+	return m.local.ConfigPath(env)
+}
+
+// Path returns the path to the environment .env file
+func (m *manager) Path(env *Environment) string {
+	return m.local.EnvPath(env)
+}
+
+// List returns a list of all environments within the data store
+func (m *manager) List(ctx context.Context) ([]*Description, error) {
+	envMap := map[string]*Description{}
+	defaultEnvName, err := m.azdContext.GetDefaultEnvironmentName()
+	if err != nil {
+		defaultEnvName = ""
+	}
+
+	localEnvs, err := m.local.List(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving local environments, %w", err)
+	}
+
+	for _, env := range localEnvs {
+		envMap[env.Name] = &Description{
+			Name:     env.Name,
+			HasLocal: true,
+		}
+	}
+
+	if m.remote != nil {
+		remoteEnvs, err := m.remote.List(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("retrieving remote environments, %w", err)
+		}
+
+		for _, env := range remoteEnvs {
+			existing, has := envMap[env.Name]
+			if !has {
+				existing = &Description{
+					Name:      env.Name,
+					HasRemote: true,
+				}
+			} else {
+				existing.HasRemote = true
+			}
+			envMap[env.Name] = existing
+		}
+
+	}
+
+	allEnvs := []*Description{}
+	for _, env := range envMap {
+		env.IsDefault = env.Name == defaultEnvName
+		allEnvs = append(allEnvs, env)
+	}
+
+	slices.SortFunc(allEnvs, func(a, b *Description) bool {
+		return a.Name < b.Name
+	})
+
+	return allEnvs, nil
+}
+
+// Get returns the environment instance for the specified environment name
+func (m *manager) Get(ctx context.Context, name string) (*Environment, error) {
+	localEnv, err := m.local.Get(ctx, name)
+	if err != nil {
+		if m.remote == nil {
+			return nil, err
+		}
+
+		remoteEnv, err := m.remote.Get(ctx, name)
```

Comment:
What if the remote copy is behind the local copy? This ends up overriding the local copy implicitly.

Reply:
The goal here is similar to terraform remote state.  The state is always synced for every state operation in azd.  Any time an environment save is performed the change is persisted locally and then immediately saved to the remote as well.  @ellismg and I spoke about this privately but agree we should document this behavior.